airbyte-cdk/python/airbyte_cdk/sources/connector_state_manager.py

```diff
@@ -40,15 +56,97 @@ def get_legacy_state(self) -> MutableMapping[str, Any]:
         """
         return copy.deepcopy(self.legacy, {})
 
-    def update_state_for_stream(self, namespace: str, stream_name: str, value: Mapping[str, Any]):
-        # todo implement in upcoming PRs
-        pass
+    def update_state_for_stream(self, stream_name: str, namespace: str, value: Mapping[str, Any]):
+        stream_descriptor = HashableStreamDescriptor(name=stream_name, namespace=namespace)
+        if stream_descriptor in self.streams:
+            self.streams[stream_descriptor].stream_state = AirbyteStateBlob.parse_obj(value)
+        else:
+            self.streams[stream_descriptor] = AirbyteStreamState(
+                stream_descriptor=StreamDescriptor(name=stream_name, namespace=namespace), stream_state=AirbyteStateBlob.parse_obj(value)
+            )
+        self.legacy[stream_name] = value
```

Comment:
yup thats a good callout, getting rid of legacy and adding additional processing to get_legacy_state, should make the internal object simpler. i'm also simplifying the map to be descriptor -> state blob (instead of descriptor -> state message)